chunk_store.go

```diff
@@ -265,39 +265,97 @@ func (c *AWSChunkStore) Put(ctx context.Context, chunks []wire.Chunk) error {
 		return err
 	}
 
-	// TODO: parallelise
+	err = c.putChunks(userID, chunks)
+	if err != nil {
+		return err
+	}
+
+	return c.updateIndex(userID, chunks)
+}
+
+// putChunks writes a collection of chunks to S3 in parallel.
+func (c *AWSChunkStore) putChunks(userID string, chunks []wire.Chunk) error {
+	incomingErrors := make(chan error)
 	for _, chunk := range chunks {
-		err := timeRequest("Put", s3RequestDuration, func() error {
-			var err error
-			_, err = c.s3.PutObject(&s3.PutObjectInput{
-				Body:   bytes.NewReader(chunk.Data),
-				Bucket: aws.String(c.bucketName),
-				Key:    aws.String(chunkName(userID, chunk.ID)),
-			})
-			return err
-		})
+		go func(chunk wire.Chunk) {
+			incomingErrors <- c.putChunk(userID, &chunk)
+		}(chunk)
+	}
+
+	var lastErr error
+	for range chunks {
+		err := <-incomingErrors
 		if err != nil {
-			return err
+			lastErr = err
 		}
+	}
+	return lastErr
```

Comment:
Yeah, it'd be nice if we could somehow "encode" this pattern in some sort of formal language. (c.f. https://jml.io/2016/08/anti-patterns.html)